### ISSUE TITLE
support rollup mode [COUNT() -> SUM(count)]

### ISSUE THREAD

**Issue by QiuMM** [apache/incubator-druid#6583]
### Motivation

In my company, we use Druid as our OLAP engine that enables business users to analyze large quantities of data in real-time. However, druid will rollup data(if enable) at ingestion time. So when our users use `count()` in their sql, they often get unexpected results. Our users are not familiar with druid, it's hard for them to write a sql that will work as expected. So it would be better that druid support rollup mode, i.e. rewrite COUNT() to SUM(count), make it transparent to users.

### Proposed solution

Add a boolean property `rollup` in `PlannerConfig` to indicate whether using the rollup mode or not. If it's true, in `CountSqlAggregator#toDruidAggregation` create a `LongSumAggregatorFactory` instead of a `CountAggregatorFactory`. 

In `DruidSchema#analysisToRowSignature`, we can get aggregators info and find the defined count metric name, then using this name to construct a `LongSumAggregatorFactory`. However, currently I could not infer which one is the count aggregator from segment metadata query result, the related issue is #6578. 

In order to support rollup mode, #6578 must be fixed first. Looking forward to your comments.

**Comment by Dylan1312** [comment]
It seems better for user's to understand that a single row may contain aggregated events rather than try to hide the optimisation. 

If you do try to hide it, you might end up with strange behaviour

E.G.

A user runs "SELECT count(field)" and gets a result of 1000
The user then runs "SELECT field" but only gets back 10 rows.

**Comment by QiuMM** [comment]
I don't agree, in my scenario, our users are not data experts and they don't care the details. Besides, it's hard to explain to everyone that a single row may contain aggregated events in druid. 

What's more, what I suggested just provided a additional option, users can decide whether to use rollup mode or not, it's a property which can be set in the query context.

**Comment by b-slim** [comment]
I don't call this 

> A user runs "SELECT count(field)" and gets a result of 1000
>The user then runs "SELECT field" but only gets back 10 rows.

a detail, it is actually correctness, am assuming your users are actually SQL fanatics thus they care about correctness, FYI that might be also the same issue with Average, once the data is rolled up you need to break SQL AVG into SUM and Sum(SUM(__count))  as well not sure if the current SQL integration is doing that.

**Comment by QiuMM** [comment]
> once the data is rolled up you need to break SQL AVG into SUM and Sum(SUM(__count)) as well not sure if the current SQL integration is doing that.

Yes, plan to do that. But before accomplish it, I need to fix #6578 in a good way.

**Comment by QiuMM** [comment]
@Dylan1312 druid support rolling up data at ingestion time, so we need to define metric columns before ingestion. However, for traditional data analysts, they just care about dimension columns, they might not even know the existence of metric columns. In most scenarios, metric columns are added by software engineers who are familiar with druid and these metrics are transparent to their users. Even if the data analysts know about the metric columns, they also have to switch their mindset between druid and other OLAP engines. So I consider it more friendly to users to let druid sql api support rollup mode.

@b-slim for #6578, do you have any ideas.

**Comment by QiuMM** [comment]
@gianm do you have any comments? I have implemented it and it worked fine in production, but the implementation wasn't perfect since the way I solved #6578 wasn't elegant.

**Comment by vogievetsky** [comment]
Just wanted to add my 2c: but I had added that feature in https://github.com/implydata/plyql (see `--rollup` flag). It was very useful but was not on by default.

**Comment by QiuMM** [comment]
@vogievetsky I known that plyql support `COUNT() -> SUM(count)`, but not everyone who use Druid would use the plyql, besides plyql is just a CLI program. IMO support this feature inside of Druid would be great.
